[sonic-utilities] Port auto negotiation CLI implementation

Junchao-Mellanox opened the pull request:
<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did

Added following CLI commands:
- config interface autoneg
- config interface advertised-speeds
- config interface type
- config interface advertised-types
- show interfaces auto-neg status

Also added unit test cases for each new CLI commands

#### How I did it

Added new subcommands to 'config interface' group. The subcommands validate the input parameter and save value to CONFIG_DB.

Added new subcommands to 'show interfaces' group. The subcommands get auto negotiation data from APPL_DB and display them.

#### How to verify it

Manual test and unit test

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

Junchao-Mellanox committed:
Add CLIs to support port auto negotiation

Junchao-Mellanox committed:
Adjust CLI for port auto negotiation

Conflicts:
	tests/mock_tables/state_db.json

Junchao-Mellanox committed:
Fix unit test issue

Junchao-Mellanox committed:
Fix unit test issues

Junchao-Mellanox committed:
Merge commit '42cab68' into port-auto-neg

Conflicts:
	scripts/portconfig
	tests/mock_tables/state_db.json

Junchao-Mellanox committed:
Change autoneg value range from 0|1 to on|off

Junchao-Mellanox committed:
Merge commit '6ced42c' into port-auto-neg

Junchao-Mellanox committed:
Fix unit test issue

Junchao-Mellanox committed:
Fixes issue: db_migrator change cause db_migrator test case failure

Junchao-Mellanox committed:
Fix issues found in manual test

Junchao-Mellanox committed:
Fix issue: db migrator not working for port auto negotiation attributes

Junchao-Mellanox committed:
Add unit test for DB migrator

Junchao-Mellanox committed:
Merge branch 'master' into port-auto-neg

keboliu commented on config/main.py:
```diff
@@ -2349,6 +2349,126 @@ def speed(ctx, interface_name, interface_speed, verbose):
+    log.log_info("'interface autoneg {} {}' executing...".format(interface_name, mode))
+
+    if ctx.obj['namespace'] is DEFAULT_NAMESPACE:
+        command = "portconfig -p {} -an {}".format(interface_name, mode)
```
-an -> --an or something else

stephenxs replied:
If we have `--auto-neg` for long, it's correct to have `-an` for short.

Junchao-Mellanox replied:
then i will keep it.

stephenxs commented on config/main.py:
```diff
@@ -2349,6 +2349,126 @@ def speed(ctx, interface_name, interface_speed, verbose):
+    if clicommon.get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(config_db, interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
```
`None` is the developer’s language. Probably `wrong interface` name is better.
Same for the following functions

Junchao-Mellanox replied:
This is following the style of other "config interface" commands.

stephenxs replied:
I still suggest changing it. This message is not clear.
The user can provide an invalid interface name which causes `interface_alias_to_name` to return `None`. But he might get confused about this message because he didn't provide an None or empty interface name.

Junchao-Mellanox replied:
I totally agree your suggestion. Probably we can have a separate PR to change this print. If we only change the print in this subcommand, it will be a litter bit weird. How do you think?

stephenxs replied:
Agree. Let's keep it as it is for now.

stephenxs commented on config/main.py:
```diff
@@ -2349,6 +2349,126 @@ def speed(ctx, interface_name, interface_speed, verbose):
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    log.log_info("'interface autoneg {} {}' executing...".format(interface_name, mode))
```
Debug? I tend to use debug here. If more information is provided, for example, get some data from db, we can use info.
Same for the following fucntions

Junchao-Mellanox replied:
This is following the style of other "config interface" commands.

stephenxs replied:
Ok. Let's keep it as it was and check whether there are comments from the community.

stephenxs commented on scripts/intfutil:
```diff
@@ -133,7 +137,13 @@ def appl_db_port_status_get(appl_db, intf_name, status_type):
+        new_speed_list = []
+        for s in natsorted(speed_list):
+            new_speed_list.append('{}G'.format(s[:-3]))
+        status = ','.join(new_speed_list)
```
Will this be too long to be displayed within one line?

Junchao-Mellanox replied:
Tested on 400G, looks OK for now. Any suggestion on shorten the display?

stephenxs commented on scripts/db_migrator.py:
```diff
@@ -473,7 +485,12 @@ def version_2_0_0(self):
         Current latest version. Nothing to do here.
         """
```
Please update this comments. It’s no longer the current last version.

stephenxs commented on scripts/db_migrator.py:
```diff
@@ -473,7 +485,12 @@ def version_2_0_0(self):
+        return 'version_2_0_1'
 
+    def version_2_0_1(self):
+        log.log_info('Handling version_2_0_1')
```
Add function header comments

stephenxs commented on scripts/intfutil:
```diff
@@ -506,10 +516,77 @@ class IntfDescription(object):
     parser = argparse.ArgumentParser(description='Display Interface information',
                                      formatter_class=argparse.RawTextHelpFormatter)
-    parser.add_argument('-c', '--command', type=str, help='get interface status or description', default=None)
+    parser.add_argument('-c', '--command', type=str, help='get interface status or description or autoneg status', default=None)
```
Suggest using full descriptions in UI. autoneg => auto negotiation status

stephenxs commented on scripts/portconfig:
```diff
@@ -54,6 +82,12 @@ class portconfig(object):
+        if supported_speeds_str:
+            if supported_speeds_str.find(str(speed)) == -1:
+                print('Invalid speed value specified: {}'.format(speed))
+                print('Valid speed value:{}'.format(supported_speeds_str))
```
values or just speeds and remove the value

stephenxs commented on scripts/portconfig:
```diff
@@ -66,6 +100,59 @@ class portconfig(object):
+            supported_speeds_str = self.get_supported_speeds(port)
+            if supported_speeds_str:
+                supported_speeds = set([x for x in supported_speeds_str.split(',')])
+                for speed in adv_speeds.split(','):
```
A better way to do this:
1. Transform the adv_speeds into a set as well
2. Let unsupported_speeds = adv_speeds_set minus supported_speeds
3. Fail if unsupported_speeds is not empty set

stephenxs commented on scripts/portconfig:
```diff
@@ -66,6 +100,59 @@ class portconfig(object):
+        if adv_interface_types != 'all':
+            interface_type_list = adv_interface_types.split(',')
+            for interface_type in interface_type_list:
+                if interface_type not in VALID_INTERFACE_TYPE_SET:
```
The same as the previous one. This can be done by using set minus operation.

stephenxs commented on scripts/portconfig:
```diff
@@ -66,6 +100,59 @@ class portconfig(object):
+            state_db = SonicV2Connector(host="127.0.0.1")
+        else:
+            state_db = SonicV2Connector(host="127.0.0.1", namesapce=namesapce, use_unix_socket_path=True)
+        state_db.connect(state_db.STATE_DB)
```
Should the state_db be stored in the object for further use?

Junchao-Mellanox replied:
No need. Each time CLI is called, a new state DB will be created.

stephenxs commented on scripts/portconfig:
```diff
@@ -66,6 +100,59 @@ class portconfig(object):
+            print("Setting interface_type %s on port %s" % (interface_type, port))
+        if interface_type not in VALID_INTERFACE_TYPE_SET:
+            print("Invalid interface type value specified: {}".format(interface_type))
+            print("Valid interface type:{}".format(','.join(VALID_INTERFACE_TYPE_SET)))
```
The value in the first sentence can be removed
In the second sentence, type => types

stephenxs commented on scripts/portconfig:
```diff
@@ -78,6 +165,14 @@ def main():
     parser.add_argument('-n', '--namespace', metavar='namespace details', type = str, required = False,
                         help = 'The asic namespace whose DB instance we need to connect', default=None)
+    parser.add_argument('-an', '--autoneg', type = str, required = False,
+                        help = 'port auto negotiation value in str', default=None)
```
Sounds like the “in str” can be removed or use “in string” instead

stephenxs commented on tests/config_an_test.py:
```diff
@@ -0,0 +1,72 @@
+        runner = CliRunner()
+        result = runner.invoke(config.config.commands["interface"].commands[command_name], para_list, obj = ctx)
+        print(result.output)
+        assert op(result.exit_code, expect_result)
```
Beside the return value, we need to verify the database as well

Junchao-Mellanox replied:
I agree. However, "config interface" command always call script portconfig, and portconfig will be a separate process. The config DB instance is created and destroyed in portconfig, I cannot get the values from it. So I did not verify the values in config DB.

stephenxs replied:
To achieve that, we can:
1. Mock the db connector at the beginning of `portconfig`, just like what has been done in `watermarkstat`
2. After that, the `portconfig` should be able to operate the database passed from `config interface` and modify it.
3. The updated database will be passed back to `config interface` and able to be checked.

But some details still need to be find out by using something like single step.

stephenxs replied:
As discussed offline, the `portconfig` runs in a different daemon from `config interface`, which causes it unable to pass the updated db back to `config interface`. So we don't have a way to verify config db for now.

stephenxs commented on tests/db_migrator_test.py:
```diff
@@ -173,8 +186,31 @@ def test_mellanox_buffer_migrator_for_warm_reboot(self, sku, topo):
+        os.environ['UTILITIES_UNIT_TESTING'] = "0"
+        dbconnector.dedicated_dbs['CONFIG_DB'] = None
+
+    def test_mellanox_buffer_migrator_negative_cold_reboot(self):
```
This is not the buffer migrator

stephenxs commented on scripts/intfutil:
```diff
@@ -506,10 +516,77 @@ class IntfDescription(object):
+                    autoneg_mode = 'enabled' if autoneg_mode == 'on' else 'disabled'
+                table.append((key,
+                              autoneg_mode,
+                              appl_db_port_status_get(self.db, key, PORT_SPEED),
```
I was thinking to fetch each field for each port from the database is not a good way. Probably we can fetch the entire table and then fetch info from the dict.

Junchao-Mellanox replied:
We cannot get the entire table for the APPL_DB.

stephenxs commented on show/interfaces/__init__.py:
```diff
@@ -465,3 +465,38 @@ def detailed(interface, period, verbose):
+#
+# autoneg group (show interfaces autoneg ...)
+#
+@interfaces.group(name='autoneg', cls=clicommon.AliasedGroup)
```
`auto-negotiation`?

Junchao-Mellanox replied:
No, this is following the HLD.

stephenxs commented on tests/intfutil_test.py:
```diff
@@ -62,6 +62,31 @@
+  Interface    Auto-Neg Mode    Speed    Adv Speeds    Type    Adv Types    Oper    Admin
+-----------  ---------------  -------  ------------  ------  -----------  ------  -------
+ Ethernet32              N/A      40G           N/A     N/A          N/A      up       up
+"""
```
Can we add some test cases in which the `Auto-Neg Mode`, `Adv Speeds` and `Adv Types` are not `N/A`

lguohan committed:
[acl_loader]: add iptype match to the rules for dataplane acl

dataplane acl has v4 and v6 type. in case the rule does not
specify the iptype, the acl_loader will automatically add
the match for the iptype based on the table type.

for l3 table, it will add ethertype = 0x800
for l3v6 table, it will add iptype = ipv6any

Signed-off-by: Guohan Lu <lguohan@gmail.com>

Junchao-Mellanox committed:
Merge commit '9a2872d' into port-auto-neg

Junchao-Mellanox committed:
Fix review comments

stephenxs commented:
You might want to rebase your PR with unrelated commits removed.

lguohan committed:
[ci]: download latest docker-sonic-vs from sonic-swss repo

ensure the pytest is running against latest sonic-swss binary

Signed-off-by: Guohan Lu <lguohan@gmail.com>

Junchao-Mellanox committed:
Fix review comments

stephenxs commented on scripts/portconfig:
```diff
@@ -114,20 +114,21 @@ class portconfig(object):
             supported_speeds_str = self.get_supported_speeds(port)
             if supported_speeds_str:
                 supported_speeds = set([x for x in supported_speeds_str.split(',')])
```
```suggestion
                supported_speeds = set(supported_speeds_str.split(','))
```

Junchao-Mellanox committed:
Fix review comment

shlomibitton committed:
[fast-reboot] Fix dump script to support PortChannels in a VLAN group (#1393)

Add PortChannels to the list of interfaces (port_id_2_iface) to support FDB dump for PortChannel in a VLAN group.

Fix Azure/sonic-buildimage#4793

- How I did it

- Get LAG ID from the DB.
- Find the LAG name from APP DB.
- Add it to the list of 'port_id_2_iface' to be used.

- How to verify it
Reproduce the issue mentioned on this PR and try to run fast-reboot with this fix.

- Previous command output (if the output of a command-line utility has changed)
Traceback:
src_ifs = {map_mac_ip_per_vlan[vlan_name][dst_mac] for vlan_name, dst_mac, _ in arp_entries}
KeyError: 'b8:59:9f:a8:e2:00'

Signed-off-by: Shlomi Bitton <shlomibi@nvidia.com>

lguohan committed:
[setup.py]: move the module to test requirement (#1542)

bug introduced in #1393

Signed-off-by: Guohan Lu <lguohan@gmail.com>

yozhao101 committed:
[reboot] No need enable the Linux magic system request key (#1538)

As introduced by the kernel doc, we do not need the line 20 in reboot script since it only influenced the capability to reboot the Linux kernel by hitting the combo magic key of keyboard.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

abdosi committed:
Reduce time taken by show commands on multi-asic platforms (#1544)

Optimize/Reduce the time taken by show interface * and show ip bgp * commands on Multi-asic platforms.

yozhao101 committed:
[config] Eliminate race condition between reloading Monit config and (#1543)

Signed-off-by: Yong Zhao yozhao@microsoft.com

What I did
Nightly test found a failure when we ran the command sudo config reload/load_minigraph, The error message is:

admin@str-a7050-acs-1:~$ sudo config reload -y
Disabling container monitoring ...
Stopping SONiC target ...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Resetting failed status on bgp.service
Resetting failed status on caclmgrd.service
Resetting failed status on dhcp_relay.service
Resetting failed status on hostcfgd.service
Resetting failed status on hostname-config.service
Resetting failed status on interfaces-config.service
Resetting failed status on lldp.service
Resetting failed status on ntp-config.service
Resetting failed status on pmon.service
Resetting failed status on procdockerstatsd.service
Resetting failed status on radv.service
Resetting failed status on rsyslog-config.service
Resetting failed status on swss.service
Resetting failed status on syncd.service
Resetting failed status on teamd.service
Resetting failed status on telemetry.timer
Restarting SONiC target ...
Reloading Monit configuration ...
Reinitializing monit daemon
Enabling container monitoring ...
Unix socket /var/run/monit.sock connection error -- No such file or directory
The root reason is that there exists an implicit race condition between the command sudo monit reload at line 701 and
the command sudo monit monitor container_checker at line 706. Both commands need access the monit.sock socket file
under the directory /var/run/.

Specifically the sudo monit reload at line 701 will re-initialize the Monit daemon, delete old monit.sock file and then create a new one. During this re-initializing process, the command sudo monit status can always execute successfully at line 704 before the old monit.sock file was deleted, but the command sudo monit monitor container_checker at line 706 will only succeed if the new monit.sock was created, otherwise it will fail and raise this error message.

How I did it
I changed the sequence between the operation to reload Monit configuration and the operation to enable monitoring container_checker.

How to verify it
I verified this change on DuT str-a7050-acs-1 by running the command sudo config reload/load_minigraph -y to make sure the error was not raised again.

Previous command output (if the output of a command-line utility has changed)
admin@str-a7050-acs-1:~$ sudo config reload -y
Disabling container monitoring ...
Stopping SONiC target ...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Resetting failed status on bgp.service
Resetting failed status on caclmgrd.service
Resetting failed status on dhcp_relay.service
Resetting failed status on hostcfgd.service
Resetting failed status on hostname-config.service
Resetting failed status on interfaces-config.service
Resetting failed status on lldp.service
Resetting failed status on ntp-config.service
Resetting failed status on pmon.service
Resetting failed status on procdockerstatsd.service
Resetting failed status on radv.service
Resetting failed status on rsyslog-config.service
Resetting failed status on swss.service
Resetting failed status on syncd.service
Resetting failed status on teamd.service
Resetting failed status on telemetry.timer
Restarting SONiC target ...
Reloading Monit configuration ...
Reinitializing monit daemon
Enabling container monitoring ...
New command output (if the output of a command-line utility has changed)
admin@str-a7050-acs-1:~$ sudo config reload -y
Disabling container monitoring ...
Stopping SONiC target ...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Resetting failed status on bgp.service
Resetting failed status on caclmgrd.service
Resetting failed status on dhcp_relay.service
Resetting failed status on hostcfgd.service
Resetting failed status on hostname-config.service
Resetting failed status on interfaces-config.service
Resetting failed status on lldp.service
Resetting failed status on ntp-config.service
Resetting failed status on pmon.service
Resetting failed status on procdockerstatsd.service
Resetting failed status on radv.service
Resetting failed status on rsyslog-config.service
Resetting failed status on swss.service
Resetting failed status on syncd.service
Resetting failed status on teamd.service
Resetting failed status on telemetry.timer
Restarting SONiC target ...
Enabling container monitoring ...
Reloading Monit configuration ...
Reinitializing monit daemon

vdahiya12 committed:
[show][config] cli support for firmware upgrade on Y-Cable (#1528)


Summary:
This PR provides the support for adding CLI commands for activatie, download, upgrade firmware on the Y-cable
In particular these Cli commands are supported:
sudo config muxcable firmware download Ethernet0
sudo config muxcable firmware rollback Ethernet0
sudo config muxcable firmware download ~/AEC_WYOMING_B52Yb0_MS_0.6_20201218.bin Ethernet0

Signed-off-by: vaibhav-dahiya vdahiya@microsoft.com

What I did
Added the support for firmware upgrade CLI on Y cable

How I did it
added the changes in sonic-utilities/show and sonic-utilities/config by changing the muxcable.py

Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

shlomibitton committed:
Fix fast-reboot-dump.py SonicV2Connector after late merge (#1546)

#### What I did
This PR https://github.com/Azure/sonic-utilities/pull/1393 merged after PR: https://github.com/Azure/sonic-utilities/pull/1392
This caused the first PR to be not aligned with SonicV2Connector change.
This PR motivation is to fix it.

lguohan committed:
[ci]: use the same snapshot of pytests as the docker-sonic-vs image (#1545)

in case sonic-swss build failure, the docker-sonic-vs may not
match exactly the same pytest checkout from master

Signed-off-by: Guohan Lu <lguohan@gmail.com>

nazariig committed:
[fwutil]: Fix next image mount (#1533)

- What I did
Fixed next image mount

- How I did it
Removed lstrip

- How to verify it
root@sonic:/home/admin# fwutil update chassis component <component_name> fw -i next

Signed-off-by: Nazarii Hnydyn <nazariig@nvidia.com>

Blueve committed:
[console] Include Flow Control status in show line result (#1549)

- Render the flow control option in show line table
- Update unit test case
- Update document

Signed-off-by: Jing Kan jika@microsoft.com

Junchao-Mellanox committed:
Merge commit '5976399' into port-auto-neg
